pkg/agent/cache/manager/manager.go

```diff
@@ -0,0 +1,302 @@
+}
+
+type EntryRequest struct {
+	CSR   []byte
+	entry cache.CacheEntry
+}
+
+type manager struct {
+	managedCache     cache.Cache
+	serverCerts      []*x509.Certificate
+	serverAddr       string
+	serverSPIFFEID   string
+	errorCh          chan error
+	entryRequestCh   chan map[string][]EntryRequest
+	regEntriesCh     chan []*proto.RegistrationEntry
+	CacheEntryCh     chan cache.CacheEntry
+	spiffeIdEntryMap map[string]cache.CacheEntry
+	baseSVID         []byte
+	baseSVIDKey      *ecdsa.PrivateKey
+	baseSPIFFEID     string
+	regEntries       []*proto.RegistrationEntry
+	log              logrus.FieldLogger
+}
+
+func NewManager(managedCache cache.Cache,
+	serverCerts []*x509.Certificate,
+	serverSPIFFEID string,
+	serverAddr string,
+	errorCh chan error,
+	baseSVID []byte, baseSVIDKey *ecdsa.PrivateKey, regEntries []*proto.RegistrationEntry, logger logrus.FieldLogger) manager {
+	cert, err := x509.ParseCertificate(baseSVID)
+	if err != nil {
+		errorCh <- err
+	}
+	basespiffeID, err := uri.GetURINamesFromCertificate(cert)
+
+	return manager{
+		managedCache:   managedCache,
+		serverCerts:    serverCerts,
+		serverSPIFFEID: serverSPIFFEID,
+		serverAddr:     serverAddr,
+		baseSVID:       baseSVID,
+		baseSVIDKey:    baseSVIDKey,
+		baseSPIFFEID:   basespiffeID[0],
+		regEntries:     regEntries,
+		errorCh:        errorCh,
+		log:            logger.WithField("subsystem_name", "cacheManager"),
+	}
+}
+
+func (m *manager) UpdateCache(stop chan struct{}) {
+	m.spiffeIdEntryMap = make(map[string]cache.CacheEntry)
+	m.entryRequestCh = make(chan map[string][]EntryRequest)
+	m.regEntriesCh = make(chan []*proto.RegistrationEntry)
+	m.errorCh = make(chan error)
+	m.CacheEntryCh = make(chan cache.CacheEntry)
+	stopRequests := make(chan struct{})
+	stopRegEntryHandler := make(chan struct{})
+	var wg sync.WaitGroup
+
+	go m.expiredCacheEntryHandler(30*time.Second, stopRequests)
+	go m.regEntriesHandler(stopRegEntryHandler)
+	m.log.Debug("Initializing Cache Manager", " baseSVIDSPIFFEId:", m.baseSPIFFEID)
+
+	m.regEntriesCh <- m.regEntries
+
+	for {
+
+		var svid []byte
+		var key *ecdsa.PrivateKey
+		select {
+		case reqs := <-m.entryRequestCh:
+			for parentId, entryRequests := range reqs {
+				wg.Add(1)
+				if _, ok := m.spiffeIdEntryMap[parentId]; ok {
+					svid = m.spiffeIdEntryMap[parentId].SVID.SvidCert
+					key = m.spiffeIdEntryMap[parentId].PrivateKey
+				}
+				if parentId == m.baseSPIFFEID {
+					svid = m.baseSVID
+					key = m.baseSVIDKey
+				}
+				conn, err := m.getGRPCConn(svid, key)
+				if err != nil {
+					m.errorCh <- err
+				}
+				m.log.Debug("Spawning FetchID for", "entryRequests:", entryRequests)
+				go m.fetchSVID(entryRequests, node.NewNodeClient(conn), &wg)
+			}
+
+		case newCacheEntry := <-m.CacheEntryCh:
+			m.managedCache.SetEntry(newCacheEntry)
+			m.log.Debug("Updated Cache", "Entry:", newCacheEntry)
+
+		case <-stop:
+			wg.Wait()
+			stopRequests <- struct{}{}
+			stopRegEntryHandler <- struct{}{}
```

Comment:
* seems like `expiredCacheEntryHandler()`, `regEntriesHandler()` should be in the wait group.
* if `stop` is closed by the parent then it will be available for reading by everything - you won't need `stopRequests` and `stopRegEntryHandler`.

Reply:
There's a bunch of race conditions where things are writing to `entryRequestCh`, `CacheEntryCh` but nothing is listening.  It won't be hard to make `wg.Wait()` never return.